### DISCUSSION TITLE
How to get flat netlist from design checkpoint generated from vivado synthesis？

### DISCUSSION THREAD

**Discussion by wengwz** [Xilinx/RapidWright#1003] [Q&A] [answered]
I'm trying to get the flat logical netlist from post-synthesis design checkpoint using RapidWright. And the flat netlist consists of only leaf/primitive cells and connections between them. However, the logical netlist is represented hierarchically in RapidWright based on the EDIF format. And I don't find any APIs in RapidWright that can helps me flat the hierarchical netlist directly. So I get the leaf cells and nets using `getAllLeafHierCellInstances` and `getPhysicalNetPinMap` and then extract flattened netlist representation using the following codes:
```java
public class FlatCircuitNetlist {

    public static void main(String[] args) {
        String dcpPath = "./benchmark/lu_20x20.dcp";
        Design circuitDesign = Design.readCheckpoint(dcpPath);
        EDIFNetlist logicalNetlist = circuitDesign.getNetlist();

        List<EDIFHierCellInst> leafHierCellInsts = logicalNetlist.getAllLeafHierCellInstances();
        Map<EDIFCellInst, EDIFHierCellInst> cell2HierCellMap = new HashMap<>();
        Map<EDIFHierNet, List<EDIFHierPortInst>> hierNet2HierPortMap = logicalNetlist.getPhysicalNetPinMap();
        Map<EDIFHierPortInst, EDIFHierNet> hierPort2HierNetMap = new HashMap<>();
        Map<EDIFHierCellInst, List<EDIFHierPortInst>> hierCell2HierPortMap = new HashMap<>();
        Map<EDIFHierPortInst, EDIFHierCellInst> hierPort2HierCellMap = new HashMap<>();

        for (EDIFHierCellInst hierCellInst : leafHierCellInsts) {
            cell2HierCellMap.put(hierCellInst.getInst(), hierCellInst);
        }

        for (Map.Entry<EDIFHierNet, List<EDIFHierPortInst>> entry : hierNet2HierPortMap.entrySet()) {
            for (EDIFHierPortInst portInst : entry.getValue()) {
                hierPort2HierNetMap.put(portInst, entry.getKey());

                EDIFHierCellInst hierCell = cell2HierCellMap.get(portInst.getPortInst().getCellInst());
                assert hierCell != null;
                
                hierPort2HierCellMap.put(portInst, hierCell);
                if (hierCell2HierPortMap.containsKey(hierCell)){
                    List<EDIFHierPortInst> portInsts = hierCell2HierPortMap.get(hierCell);
                    portInsts.add(portInst);
                } else {
                    List<EDIFHierPortInst> portInsts = new ArrayList<>();
                    portInsts.add(portInst);
                    hierCell2HierPortMap.put(hierCell, portInsts);
                }
            }
        }
        System.out.println("Generate Flat Netlist Successfully");
    }
}
```
Can anyone help me verify if this implementation is correct and capable of extracting the complete information of the flat netlist? Specifically, I want to know if `getPhysicalNetPinMap` function returns all connections information between leaf cells. And are there more convenient methods or APIs to extract a flat netlist that includes only leaf cells and their connections?

Besides, I find that this implementation doesn't work for netlists containing blockbox cells. Specifically, it seems that `getAllLeafHierCellInstances` can't extract the blockbox cells as leaf cell and when I want to map PortInst of blackbox cell to their corresponding EDIFHierCellInst, the assertion `assert hierCell != null` fails. So how can I extract all blackbox EDIFCellInst and their corresponding EDIFHierCellInst?

**Comment by clavin-xlnx** [comment, marked as answer]
The easiest approach is to create a new flattened netlist as a copy of the existing hierarchical one.   Trying to modify an existing netlist in place to remove all the hierarchy could be done, but it is probably better to avoid that approach.  I'm not sure about the code above, it is touching on some of the important APIs that are needed, but it doesn't address some of the corner cases such as VCC and GND nets and their respective sources that need to be collapsed in a flattened netlist.  

 I've drafted up a PR that includes code that can accomplish this, see #1006 for details, it includes a test that should demonstrate how it can be used and will accommodate designs that contain black boxes.

**Reply by wengwz** [reply]
Your response and code have perfectly solved the problems I was facing. Thank you so much!

**Reply by wengwz** [reply]
Hi Chris, I have some questions regarding the details of #1006. Why do we need to call the function flatNetlist.expandMacroUnisims(part.getSeries()) to restore the hierarchy of Unisim macros in the flat netlist? Does this means that we can't separate the leaf cells of Unisim macros and that they must be handled as a whole during the subsequent placement and routing phases?

**Reply by clavin-xlnx** [reply]
At the beginning of the method, we collapse the macros since macros are leaf cells that we want to preserve.  To be consistent, we expand the macros again at the end of the method.  This is to maintain integrity of the macros after the design is flattened.  Most macros, if flattened, might lead to an invalid netlist so we need to maintain their existence.
